[C,php/elFinderVolumeDriver.class.php,3036] 	}
[C,php/elFinderVolumeDriver.class.php,3037] 	
[A,php/elFinderVolumeDriver.class.php,3038] 	/**
[A,php/elFinderVolumeDriver.class.php,3039] 	 * Get temporary contents link infomation
[A,php/elFinderVolumeDriver.class.php,3040] 	 * 
[A,php/elFinderVolumeDriver.class.php,3041] 	 * @param string $name
[A,php/elFinderVolumeDriver.class.php,3042] 	 * @return boolean|array
[A,php/elFinderVolumeDriver.class.php,3043] 	 * @author Naoki Sawada
[A,php/elFinderVolumeDriver.class.php,3044] 	 */
[A,php/elFinderVolumeDriver.class.php,3045] 	public function getTempLinkInfo($name = null) {
[A,php/elFinderVolumeDriver.class.php,3046] 		if ($this->tmpLinkPath) {
[A,php/elFinderVolumeDriver.class.php,3047] 			if (! $name) {
[A,php/elFinderVolumeDriver.class.php,3048] 				$name = 'temp_' . md5($_SERVER['REMOTE_ADDR'] . (string)microtime(true));
[A,php/elFinderVolumeDriver.class.php,3049] 			} else if (substr($name, 0, 5) !== 'temp_') {
[A,php/elFinderVolumeDriver.class.php,3050] 				$name = 'temp_' . $name;
[A,php/elFinderVolumeDriver.class.php,3051] 			}
[A,php/elFinderVolumeDriver.class.php,3052] 			register_shutdown_function(array('elFinder', 'GlobGC'), $this->tmpLinkPath . DIRECTORY_SEPARATOR . 'temp_*', elFinder::$tmpLinkLifeTime);
[A,php/elFinderVolumeDriver.class.php,3053] 			return array(
[A,php/elFinderVolumeDriver.class.php,3054] 				'path' => $path = $this->tmpLinkPath . DIRECTORY_SEPARATOR . $name,
[A,php/elFinderVolumeDriver.class.php,3055] 				'url'  => $this->tmpLinkUrl . '/' . $name
[A,php/elFinderVolumeDriver.class.php,3056] 			);
[A,php/elFinderVolumeDriver.class.php,3057] 		}
[A,php/elFinderVolumeDriver.class.php,3058] 		return false;
[A,php/elFinderVolumeDriver.class.php,3059] 	}
[A,php/elFinderVolumeDriver.class.php,3060] 	
[A,php/elFinderVolumeDriver.class.php,3061] 	/**
[A,php/elFinderVolumeDriver.class.php,3062] 	 * Get URL of substitute image by request args `substitute` or 4th argument $maxSize
[A,php/elFinderVolumeDriver.class.php,3063] 	 * 
[A,php/elFinderVolumeDriver.class.php,3064] 	 * @param string   $target  Target hash
[A,php/elFinderVolumeDriver.class.php,3065] 	 * @param array    $srcSize Size info array [width, height]
[A,php/elFinderVolumeDriver.class.php,3066] 	 * @param resource $srcfp   Source file file pointer
[A,php/elFinderVolumeDriver.class.php,3067] 	 * @param integer  $maxSize Maximum pixel of substitute image
[A,php/elFinderVolumeDriver.class.php,3068] 	 * @return boolean
[A,php/elFinderVolumeDriver.class.php,3069] 	 */
[A,php/elFinderVolumeDriver.class.php,3070] 	public function getSubstituteImgLink($target, $srcSize, $srcfp = null, $maxSize = null) {
[A,php/elFinderVolumeDriver.class.php,3071] 		$url = false;
[A,php/elFinderVolumeDriver.class.php,3072] 		if (! $maxSize) {
[A,php/elFinderVolumeDriver.class.php,3073] 			$args = elFinder::$currentArgs;
[A,php/elFinderVolumeDriver.class.php,3074] 			$maxSize = $args['substitute'];
[A,php/elFinderVolumeDriver.class.php,3075] 		}
[A,php/elFinderVolumeDriver.class.php,3076] 		if ($maxSize) {
[A,php/elFinderVolumeDriver.class.php,3077] 			if ($this->getOption('substituteImg')) {
[A,php/elFinderVolumeDriver.class.php,3078] 				$maxSize = intval($maxSize);
[A,php/elFinderVolumeDriver.class.php,3079] 				$zoom = min(($maxSize/$srcSize[0]),($maxSize/$srcSize[1]));
[A,php/elFinderVolumeDriver.class.php,3080] 				if ($zoom < 1) {
[A,php/elFinderVolumeDriver.class.php,3081] 					$width = round($srcSize[0] * $zoom);
[A,php/elFinderVolumeDriver.class.php,3082] 					$height = round($srcSize[1] * $zoom);
[A,php/elFinderVolumeDriver.class.php,3083] 					$jpgQuality = 50;
[A,php/elFinderVolumeDriver.class.php,3084] 					$preserveExif = false;
[A,php/elFinderVolumeDriver.class.php,3085] 					$unenlarge = true;
[A,php/elFinderVolumeDriver.class.php,3086] 					$checkAnimated = true;
[A,php/elFinderVolumeDriver.class.php,3087] 					if (! $srcfp) {
[A,php/elFinderVolumeDriver.class.php,3088] 						elFinder::checkAborted();
[A,php/elFinderVolumeDriver.class.php,3089] 						$srcfp = $this->open($target);
[A,php/elFinderVolumeDriver.class.php,3090] 					}
[A,php/elFinderVolumeDriver.class.php,3091] 					if ($srcfp && ($tempLink = $this->getTempLinkInfo())) {
[A,php/elFinderVolumeDriver.class.php,3092] 						elFinder::checkAborted();
[A,php/elFinderVolumeDriver.class.php,3093] 						$dest = fopen($tempLink['path'], 'wb');
[A,php/elFinderVolumeDriver.class.php,3094] 						if ($dest && stream_copy_to_stream($srcfp, $dest)) {
[A,php/elFinderVolumeDriver.class.php,3095] 							fclose($dest);
[A,php/elFinderVolumeDriver.class.php,3096] 							if ($this->imageUtil('resize', $tempLink['path'], compact('width', 'height', 'jpgQuality', 'preserveExif', 'unenlarge', 'checkAnimated'))) {
[A,php/elFinderVolumeDriver.class.php,3097] 								$url = $tempLink['url'];
[A,php/elFinderVolumeDriver.class.php,3098] 								// set expire to 1 min left
[A,php/elFinderVolumeDriver.class.php,3099] 								touch($tempLink['path'], time() - elFinder::$tmpLinkLifeTime + 60);
[A,php/elFinderVolumeDriver.class.php,3100] 							} else {
[A,php/elFinderVolumeDriver.class.php,3101] 								unlink($tempLink['path']);
[A,php/elFinderVolumeDriver.class.php,3102] 							}
[A,php/elFinderVolumeDriver.class.php,3103] 						}
[A,php/elFinderVolumeDriver.class.php,3104] 						$this->close($srcfp, $target);
[A,php/elFinderVolumeDriver.class.php,3105] 					}
[A,php/elFinderVolumeDriver.class.php,3106] 				}
[A,php/elFinderVolumeDriver.class.php,3107] 			}
[A,php/elFinderVolumeDriver.class.php,3108] 		}
[A,php/elFinderVolumeDriver.class.php,3109] 		
[A,php/elFinderVolumeDriver.class.php,3110] 		return $url;
[A,php/elFinderVolumeDriver.class.php,3111] 	}
[A,php/elFinderVolumeDriver.class.php,3112] 	
[C,php/elFinderVolumeDriver.class.php,3113] 	/**
[C,php/elFinderVolumeDriver.class.php,3114] 	 * Return temp path
[C,php/elFinderVolumeDriver.class.php,3115] 	 * 
[C,php/elFinderVolumeDriver.class.php,3116] 	 * @return string
[C,php/elFinderVolumeDriver.class.php,3117] 	 * @author Naoki Sawada
[C,php/elFinderVolumeDriver.class.php,3118] 	 */
[C,php/elFinderVolumeDriver.class.php,3119] 	public function getTempPath() {
[C,php/elFinderVolumeDriver.class.php,3120] 		$tempPath = null;
[C,php/elFinderVolumeDriver.class.php,3121] 		if (isset($this->tmpPath) && $this->tmpPath && is_writable($this->tmpPath)) {
[C,php/elFinderVolumeDriver.class.php,3122] 			$tempPath = $this->tmpPath;
[C,php/elFinderVolumeDriver.class.php,3123] 		} else if (isset($this->tmp) && $this->tmp && is_writable($this->tmp)) {
[C,php/elFinderVolumeDriver.class.php,3124] 			$tempPath = $this->tmp;
[C,php/elFinderVolumeDriver.class.php,3125] 		} else if (elFinder::getStaticVar('commonTempPath') && is_writable(elFinder::getStaticVar('commonTempPath'))) {
[C,php/elFinderVolumeDriver.class.php,3126] 			$tempPath = elFinder::getStaticVar('commonTempPath');
[C,php/elFinderVolumeDriver.class.php,3127] 		} else if (function_exists('sys_get_temp_dir')) {
@@ -3711,47 +3794,51 @@ abstract class elFinderVolumeDriver {
[C,php/elFinderVolumeDriver.class.php,3794] 	}
[C,php/elFinderVolumeDriver.class.php,3795] 	
[C,php/elFinderVolumeDriver.class.php,3796] 	/**
[C,php/elFinderVolumeDriver.class.php,3797] 	 * Get image size array with `dimensions`
[C,php/elFinderVolumeDriver.class.php,3798] 	 *
[C,php/elFinderVolumeDriver.class.php,3799] 	 * @param string $path path need convert encoding to server encoding
[C,php/elFinderVolumeDriver.class.php,3800] 	 * @param string $mime file mime type
[C,php/elFinderVolumeDriver.class.php,3801] 	 * @return array|false
[C,php/elFinderVolumeDriver.class.php,3802] 	 */
[C,php/elFinderVolumeDriver.class.php,3803] 	public function getImageSize($path, $mime = '') {
[C,php/elFinderVolumeDriver.class.php,3804] 		$size = false;
[C,php/elFinderVolumeDriver.class.php,3805] 		if ($mime === '' || strtolower(substr($mime, 0, 5)) === 'image') {
[C,php/elFinderVolumeDriver.class.php,3806] 			if ($work = $this->getWorkFile($path)) {
[C,php/elFinderVolumeDriver.class.php,3807] 				if ($size = getimagesize($work)) {
[C,php/elFinderVolumeDriver.class.php,3808] 					$size['dimensions'] = $size[0].'x'.$size[1];
[A,php/elFinderVolumeDriver.class.php,3809] 					$srcfp = fopen($work, 'rb');
[A,php/elFinderVolumeDriver.class.php,3810] 					if ($subImgLink = $this->getSubstituteImgLink(elFinder::$currentArgs['target'], $size, $srcfp)) {
[A,php/elFinderVolumeDriver.class.php,3811] 						$size['url'] = $subImgLink;
[A,php/elFinderVolumeDriver.class.php,3812] 					}
[C,php/elFinderVolumeDriver.class.php,3813] 				}
[C,php/elFinderVolumeDriver.class.php,3814] 			}
[C,php/elFinderVolumeDriver.class.php,3815] 			is_file($work) && unlink($work);
[C,php/elFinderVolumeDriver.class.php,3816] 		}
[C,php/elFinderVolumeDriver.class.php,3817] 		return $size;
[C,php/elFinderVolumeDriver.class.php,3818] 	}
[C,php/elFinderVolumeDriver.class.php,3819] 	
[C,php/elFinderVolumeDriver.class.php,3820] 	/**
[C,php/elFinderVolumeDriver.class.php,3821] 	 * Delete dirctory trees
[C,php/elFinderVolumeDriver.class.php,3822] 	 *
[C,php/elFinderVolumeDriver.class.php,3823] 	 * @param string $localpath path need convert encoding to server encoding
[C,php/elFinderVolumeDriver.class.php,3824] 	 * @return boolean
[C,php/elFinderVolumeDriver.class.php,3825] 	 * @author Naoki Sawada
[C,php/elFinderVolumeDriver.class.php,3826] 	 */
[C,php/elFinderVolumeDriver.class.php,3827] 	protected function delTree($localpath) {
[C,php/elFinderVolumeDriver.class.php,3828] 		foreach ($this->_scandir($localpath) as $p) {
[D,php/elFinderVolumeDriver.class.php,3742] 			elFinder::extendTimeLimit();
[A,php/elFinderVolumeDriver.class.php,3829] 			elFinder::checkAborted();
[C,php/elFinderVolumeDriver.class.php,3830] 			$stat = $this->stat($this->convEncOut($p));
[C,php/elFinderVolumeDriver.class.php,3831] 			$this->convEncIn();
[C,php/elFinderVolumeDriver.class.php,3832] 			($stat['mime'] === 'directory')? $this->delTree($p) : $this->_unlink($p);
[C,php/elFinderVolumeDriver.class.php,3833] 		}
[C,php/elFinderVolumeDriver.class.php,3834] 		$res = $this->_rmdir($localpath);
[C,php/elFinderVolumeDriver.class.php,3835] 		$res && $this->clearstatcache();
[C,php/elFinderVolumeDriver.class.php,3836] 		return $res;
[C,php/elFinderVolumeDriver.class.php,3837] 	}
[C,php/elFinderVolumeDriver.class.php,3838] 	
[C,php/elFinderVolumeDriver.class.php,3839] 	/**
[C,php/elFinderVolumeDriver.class.php,3840] 	 * Copy items to a new temporary directory on the local server
[C,php/elFinderVolumeDriver.class.php,3841] 	 * 
[C,php/elFinderVolumeDriver.class.php,3842] 	 * @param  array  $hashes  target hashes
[C,php/elFinderVolumeDriver.class.php,3843] 	 * @param  string $dir     destination directory (for recurcive)
[C,php/elFinderVolumeDriver.class.php,3844] 	 * @param  string $canLink it can use link() (for recurcive)
@@ -3762,31 +3849,31 @@ abstract class elFinderVolumeDriver {
[C,php/elFinderVolumeDriver.class.php,3849] 		static $totalSize = 0;
[C,php/elFinderVolumeDriver.class.php,3850] 		if (is_null($dir)) {
[C,php/elFinderVolumeDriver.class.php,3851] 			$totalSize = 0;
[C,php/elFinderVolumeDriver.class.php,3852] 			if (! $tmpDir = $this->getTempPath()) {
[C,php/elFinderVolumeDriver.class.php,3853] 				return false;
[C,php/elFinderVolumeDriver.class.php,3854] 			}
[C,php/elFinderVolumeDriver.class.php,3855] 			$dir = tempnam($tmpDir, 'elf');
[C,php/elFinderVolumeDriver.class.php,3856] 			if (!unlink($dir) || !mkdir($dir, 0700, true)) {
[C,php/elFinderVolumeDriver.class.php,3857] 				return false;
[C,php/elFinderVolumeDriver.class.php,3858] 			}
[C,php/elFinderVolumeDriver.class.php,3859] 			register_shutdown_function(array($this, 'rmdirRecursive'), $dir);
[C,php/elFinderVolumeDriver.class.php,3860] 		}
[C,php/elFinderVolumeDriver.class.php,3861] 		if (is_null($canLink)) {
[C,php/elFinderVolumeDriver.class.php,3862] 			$canLink = ($this instanceof elFinderVolumeLocalFileSystem);
[C,php/elFinderVolumeDriver.class.php,3863] 		}
[D,php/elFinderVolumeDriver.class.php,3777] 		elFinder::extendTimeLimit();
[A,php/elFinderVolumeDriver.class.php,3864] 		elFinder::checkAborted();
[C,php/elFinderVolumeDriver.class.php,3865] 		$res = true;
[C,php/elFinderVolumeDriver.class.php,3866] 		$files = array();
[C,php/elFinderVolumeDriver.class.php,3867] 		foreach ($hashes as $hash) {
[C,php/elFinderVolumeDriver.class.php,3868] 			if (($file = $this->file($hash)) == false) {
[C,php/elFinderVolumeDriver.class.php,3869] 				continue;
[C,php/elFinderVolumeDriver.class.php,3870] 			}
[C,php/elFinderVolumeDriver.class.php,3871] 			if (!$file['read']) {
[C,php/elFinderVolumeDriver.class.php,3872] 				continue;
[C,php/elFinderVolumeDriver.class.php,3873] 			}
[C,php/elFinderVolumeDriver.class.php,3874] 			
[C,php/elFinderVolumeDriver.class.php,3875] 			$name = $file['name'];
[C,php/elFinderVolumeDriver.class.php,3876] 			// for call from search results
[C,php/elFinderVolumeDriver.class.php,3877] 			if (isset($files[$name])) {
[C,php/elFinderVolumeDriver.class.php,3878] 				$name = preg_replace('/^(.*?)(\..*)?$/', '$1_'.$files[$name]++.'$2', $name);
[C,php/elFinderVolumeDriver.class.php,3879] 			} else {
@@ -4281,31 +4368,31 @@ abstract class elFinderVolumeDriver {
[C,php/elFinderVolumeDriver.class.php,4368] 			$pinfo = pathinfo($path); 
[C,php/elFinderVolumeDriver.class.php,4369] 			$ext   = isset($pinfo['extension']) ? strtolower($pinfo['extension']) : '';
[C,php/elFinderVolumeDriver.class.php,4370] 		}
[C,php/elFinderVolumeDriver.class.php,4371] 		return ($ext && isset(elFinderVolumeDriver::$mimetypes[$ext])) ? elFinderVolumeDriver::$mimetypes[$ext] : 'unknown';
[C,php/elFinderVolumeDriver.class.php,4372] 	}
[C,php/elFinderVolumeDriver.class.php,4373] 	
[C,php/elFinderVolumeDriver.class.php,4374] 	/**
[C,php/elFinderVolumeDriver.class.php,4375] 	 * Return file/total directory size
[C,php/elFinderVolumeDriver.class.php,4376] 	 *
[C,php/elFinderVolumeDriver.class.php,4377] 	 * @param  string  $path  file path
[C,php/elFinderVolumeDriver.class.php,4378] 	 * @return int
[C,php/elFinderVolumeDriver.class.php,4379] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeDriver.class.php,4380] 	 **/
[C,php/elFinderVolumeDriver.class.php,4381] 	protected function countSize($path) {
[C,php/elFinderVolumeDriver.class.php,4382] 		
[D,php/elFinderVolumeDriver.class.php,4296] 		elFinder::extendTimeLimit();
[A,php/elFinderVolumeDriver.class.php,4383] 		elFinder::checkAborted();
[C,php/elFinderVolumeDriver.class.php,4384] 		
[C,php/elFinderVolumeDriver.class.php,4385] 		$result = array('size' => 0, 'files' => 0, 'dirs' => 0);
[C,php/elFinderVolumeDriver.class.php,4386] 		$stat = $this->stat($path);
[C,php/elFinderVolumeDriver.class.php,4387] 
[C,php/elFinderVolumeDriver.class.php,4388] 		if (empty($stat) || !$stat['read'] || !empty($stat['hidden'])) {
[C,php/elFinderVolumeDriver.class.php,4389] 			$result['size'] = 'unknown';
[C,php/elFinderVolumeDriver.class.php,4390] 			return $result;
[C,php/elFinderVolumeDriver.class.php,4391] 		}
[C,php/elFinderVolumeDriver.class.php,4392] 		
[C,php/elFinderVolumeDriver.class.php,4393] 		if ($stat['mime'] !== 'directory') {
[C,php/elFinderVolumeDriver.class.php,4394] 			$result['size'] = intval($stat['size']);
[C,php/elFinderVolumeDriver.class.php,4395] 			$result['files'] = 1;
[C,php/elFinderVolumeDriver.class.php,4396] 			return $result;
[C,php/elFinderVolumeDriver.class.php,4397] 		}
[C,php/elFinderVolumeDriver.class.php,4398] 		
@@ -4566,31 +4653,31 @@ abstract class elFinderVolumeDriver {
[C,php/elFinderVolumeDriver.class.php,4653] 		
[C,php/elFinderVolumeDriver.class.php,4654] 	/**********************  manuipulations  ******************/
[C,php/elFinderVolumeDriver.class.php,4655] 		
[C,php/elFinderVolumeDriver.class.php,4656] 	/**
[C,php/elFinderVolumeDriver.class.php,4657] 	 * Copy file/recursive copy dir only in current volume.
[C,php/elFinderVolumeDriver.class.php,4658] 	 * Return new file path or false.
[C,php/elFinderVolumeDriver.class.php,4659] 	 *
[C,php/elFinderVolumeDriver.class.php,4660] 	 * @param  string  $src   source path
[C,php/elFinderVolumeDriver.class.php,4661] 	 * @param  string  $dst   destination dir path
[C,php/elFinderVolumeDriver.class.php,4662] 	 * @param  string  $name  new file name (optionaly)
[C,php/elFinderVolumeDriver.class.php,4663] 	 * @return string|false
[C,php/elFinderVolumeDriver.class.php,4664] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeDriver.class.php,4665] 	 **/
[C,php/elFinderVolumeDriver.class.php,4666] 	protected function copy($src, $dst, $name) {
[C,php/elFinderVolumeDriver.class.php,4667] 		
[D,php/elFinderVolumeDriver.class.php,4581] 		elFinder::extendTimeLimit();
[A,php/elFinderVolumeDriver.class.php,4668] 		elFinder::checkAborted();
[C,php/elFinderVolumeDriver.class.php,4669] 		
[C,php/elFinderVolumeDriver.class.php,4670] 		$srcStat = $this->stat($src);
[C,php/elFinderVolumeDriver.class.php,4671] 		
[C,php/elFinderVolumeDriver.class.php,4672] 		if (!empty($srcStat['thash'])) {
[C,php/elFinderVolumeDriver.class.php,4673] 			$target = $this->decode($srcStat['thash']);
[C,php/elFinderVolumeDriver.class.php,4674] 			if (!$this->inpathCE($target, $this->root)) {
[C,php/elFinderVolumeDriver.class.php,4675] 				return $this->setError(elFinder::ERROR_COPY, $this->path($srcStat['hash']), elFinder::ERROR_MKOUTLINK);
[C,php/elFinderVolumeDriver.class.php,4676] 			}
[C,php/elFinderVolumeDriver.class.php,4677] 			$stat   = $this->stat($target);
[C,php/elFinderVolumeDriver.class.php,4678] 			$this->clearcache();
[C,php/elFinderVolumeDriver.class.php,4679] 			return $stat && $this->symlinkCE($target, $dst, $name)
[C,php/elFinderVolumeDriver.class.php,4680] 				? $this->joinPathCE($dst, $name)
[C,php/elFinderVolumeDriver.class.php,4681] 				: $this->setError(elFinder::ERROR_COPY, $this->path($srcStat['hash']));
[C,php/elFinderVolumeDriver.class.php,4682] 		} 
[C,php/elFinderVolumeDriver.class.php,4683] 		
@@ -4669,31 +4756,31 @@ abstract class elFinderVolumeDriver {
[C,php/elFinderVolumeDriver.class.php,4756] 	}
[C,php/elFinderVolumeDriver.class.php,4757] 
[C,php/elFinderVolumeDriver.class.php,4758] 	/**
[C,php/elFinderVolumeDriver.class.php,4759] 	 * Copy file from another volume.
[C,php/elFinderVolumeDriver.class.php,4760] 	 * Return new file path or false.
[C,php/elFinderVolumeDriver.class.php,4761] 	 *
[C,php/elFinderVolumeDriver.class.php,4762] 	 * @param  Object  $volume       source volume
[C,php/elFinderVolumeDriver.class.php,4763] 	 * @param  string  $src          source file hash
[C,php/elFinderVolumeDriver.class.php,4764] 	 * @param  string  $destination  destination dir path
[C,php/elFinderVolumeDriver.class.php,4765] 	 * @param  string  $name         file name
[C,php/elFinderVolumeDriver.class.php,4766] 	 * @return string|false
[C,php/elFinderVolumeDriver.class.php,4767] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeDriver.class.php,4768] 	 **/
[C,php/elFinderVolumeDriver.class.php,4769] 	protected function copyFrom($volume, $src, $destination, $name) {
[C,php/elFinderVolumeDriver.class.php,4770] 		
[D,php/elFinderVolumeDriver.class.php,4684] 		elFinder::extendTimeLimit();
[A,php/elFinderVolumeDriver.class.php,4771] 		elFinder::checkAborted();
[C,php/elFinderVolumeDriver.class.php,4772] 		
[C,php/elFinderVolumeDriver.class.php,4773] 		if (($source = $volume->file($src)) == false) {
[C,php/elFinderVolumeDriver.class.php,4774] 			return $this->addError(elFinder::ERROR_COPY, '#'.$src, $volume->error());
[C,php/elFinderVolumeDriver.class.php,4775] 		}
[C,php/elFinderVolumeDriver.class.php,4776] 		
[C,php/elFinderVolumeDriver.class.php,4777] 		$srcIsDir = ($source['mime'] === 'directory');
[C,php/elFinderVolumeDriver.class.php,4778] 		
[C,php/elFinderVolumeDriver.class.php,4779] 		$errpath = $volume->path($source['hash']);
[C,php/elFinderVolumeDriver.class.php,4780] 		
[C,php/elFinderVolumeDriver.class.php,4781] 		if (!$this->nameAccepted($source['name'], $srcIsDir)) {
[C,php/elFinderVolumeDriver.class.php,4782] 			return $this->addError(elFinder::ERROR_COPY, $errpath, $srcIsDir? elFinder::ERROR_INVALID_DIRNAME : elFinder::ERROR_INVALID_NAME);
[C,php/elFinderVolumeDriver.class.php,4783] 		}
[C,php/elFinderVolumeDriver.class.php,4784] 		
[C,php/elFinderVolumeDriver.class.php,4785] 		if (!$source['read']) {
[C,php/elFinderVolumeDriver.class.php,4786] 			return $this->addError(elFinder::ERROR_COPY, $errpath, elFinder::ERROR_PERM_DENIED);
@@ -4714,33 +4801,38 @@ abstract class elFinderVolumeDriver {
[C,php/elFinderVolumeDriver.class.php,4801] 				$this->copyFrom($volume, $entr['hash'], $path, $entr['name']);
[C,php/elFinderVolumeDriver.class.php,4802] 			}
[C,php/elFinderVolumeDriver.class.php,4803] 			
[C,php/elFinderVolumeDriver.class.php,4804] 			$this->added[] = $test;
[C,php/elFinderVolumeDriver.class.php,4805] 		} else {
[C,php/elFinderVolumeDriver.class.php,4806] 			// MIME check
[C,php/elFinderVolumeDriver.class.php,4807] 			$mimeByName = $this->mimetype($source['name'], true);
[C,php/elFinderVolumeDriver.class.php,4808] 			if ($source['mime'] === $mimeByName) {
[C,php/elFinderVolumeDriver.class.php,4809] 				$mimeByName = '';
[C,php/elFinderVolumeDriver.class.php,4810] 			}
[C,php/elFinderVolumeDriver.class.php,4811] 			if (!$this->allowPutMime($source['mime']) || ($mimeByName && !$this->allowPutMime($mimeByName))) {
[C,php/elFinderVolumeDriver.class.php,4812] 				return $this->addError(elFinder::ERROR_UPLOAD_FILE_MIME, $errpath);
[C,php/elFinderVolumeDriver.class.php,4813] 			}
[C,php/elFinderVolumeDriver.class.php,4814] 			
[C,php/elFinderVolumeDriver.class.php,4815] 			if (strpos($source['mime'], 'image') === 0 && ($dim = $volume->dimensions($src))) {
[D,php/elFinderVolumeDriver.class.php,4729] 				$s = explode('x', $dim);
[D,php/elFinderVolumeDriver.class.php,4730] 				$source['width']  = $s[0];
[D,php/elFinderVolumeDriver.class.php,4731] 				$source['height'] = $s[1];
[A,php/elFinderVolumeDriver.class.php,4816] 				if (is_array($dim)) {
[A,php/elFinderVolumeDriver.class.php,4817] 					$dim = isset($dim['dim'])? $dim['dim'] : null;
[A,php/elFinderVolumeDriver.class.php,4818] 				}
[A,php/elFinderVolumeDriver.class.php,4819] 				if ($dim) {
[A,php/elFinderVolumeDriver.class.php,4820] 					$s = explode('x', $dim);
[A,php/elFinderVolumeDriver.class.php,4821] 					$source['width']  = $s[0];
[A,php/elFinderVolumeDriver.class.php,4822] 					$source['height'] = $s[1];
[A,php/elFinderVolumeDriver.class.php,4823] 				}
[C,php/elFinderVolumeDriver.class.php,4824] 			}
[C,php/elFinderVolumeDriver.class.php,4825] 			
[C,php/elFinderVolumeDriver.class.php,4826] 			if (($fp = $volume->open($src)) == false
[C,php/elFinderVolumeDriver.class.php,4827] 			|| ($path = $this->saveCE($fp, $destination, $name, $source)) == false) {
[C,php/elFinderVolumeDriver.class.php,4828] 				$fp && $volume->close($fp, $src);
[C,php/elFinderVolumeDriver.class.php,4829] 				return $this->addError(elFinder::ERROR_COPY, $errpath);
[C,php/elFinderVolumeDriver.class.php,4830] 			}
[C,php/elFinderVolumeDriver.class.php,4831] 			$volume->close($fp, $src);
[C,php/elFinderVolumeDriver.class.php,4832] 
[C,php/elFinderVolumeDriver.class.php,4833] 			$this->added[] = $this->stat($path);;
[C,php/elFinderVolumeDriver.class.php,4834] 		}
[C,php/elFinderVolumeDriver.class.php,4835] 		
[C,php/elFinderVolumeDriver.class.php,4836] 		return $path;
[C,php/elFinderVolumeDriver.class.php,4837] 	}
[C,php/elFinderVolumeDriver.class.php,4838] 		
@@ -1097,31 +1097,35 @@ class elFinderVolumeDropbox2 extends elFinderVolumeDriver
[C,php/elFinderVolumeDropbox2.class.php,1097]      *
[C,php/elFinderVolumeDropbox2.class.php,1098]      * @author Naoki Sawada
[C,php/elFinderVolumeDropbox2.class.php,1099]      **/
[C,php/elFinderVolumeDropbox2.class.php,1100]     protected function _dimensions($path, $mime)
[C,php/elFinderVolumeDropbox2.class.php,1101]     {
[C,php/elFinderVolumeDropbox2.class.php,1102]         if (strpos($mime, 'image') !== 0) {
[C,php/elFinderVolumeDropbox2.class.php,1103]             return '';
[C,php/elFinderVolumeDropbox2.class.php,1104]         }
[C,php/elFinderVolumeDropbox2.class.php,1105]         $ret = '';
[C,php/elFinderVolumeDropbox2.class.php,1106] 
[C,php/elFinderVolumeDropbox2.class.php,1107]         if ($data = $this->_getContents($path)) {
[C,php/elFinderVolumeDropbox2.class.php,1108]             $tmp = $this->getTempFile();
[C,php/elFinderVolumeDropbox2.class.php,1109]             file_put_contents($tmp, $data);
[C,php/elFinderVolumeDropbox2.class.php,1110]             $size = getimagesize($tmp);
[C,php/elFinderVolumeDropbox2.class.php,1111]             if ($size) {
[D,php/elFinderVolumeDropbox2.class.php,1112]                 return $size[0].'x'.$size[1];
[A,php/elFinderVolumeDropbox2.class.php,1112]                 $ret = array('dim' => $size[0].'x'.$size[1]);
[A,php/elFinderVolumeDropbox2.class.php,1113]                 $srcfp = fopen($tmp, 'rb');
[A,php/elFinderVolumeDropbox2.class.php,1114]                 if ($subImgLink = $this->getSubstituteImgLink(elFinder::$currentArgs['target'], $size, $srcfp)) {
[A,php/elFinderVolumeDropbox2.class.php,1115]                 	$ret['url'] = $subImgLink;
[A,php/elFinderVolumeDropbox2.class.php,1116]                 }
[C,php/elFinderVolumeDropbox2.class.php,1117]             }
[C,php/elFinderVolumeDropbox2.class.php,1118]         }
[C,php/elFinderVolumeDropbox2.class.php,1119] 
[C,php/elFinderVolumeDropbox2.class.php,1120]         return $ret;
[C,php/elFinderVolumeDropbox2.class.php,1121]     }
[C,php/elFinderVolumeDropbox2.class.php,1122] 
[C,php/elFinderVolumeDropbox2.class.php,1123]     /******************** file/dir content *********************/
[C,php/elFinderVolumeDropbox2.class.php,1124] 
[C,php/elFinderVolumeDropbox2.class.php,1125]     /**
[C,php/elFinderVolumeDropbox2.class.php,1126]      * Return files list in directory.
[C,php/elFinderVolumeDropbox2.class.php,1127]      *
[C,php/elFinderVolumeDropbox2.class.php,1128]      * @param string $path dir path
[C,php/elFinderVolumeDropbox2.class.php,1129]      *
[C,php/elFinderVolumeDropbox2.class.php,1130]      * @return array
[C,php/elFinderVolumeDropbox2.class.php,1131]      *
@@ -972,31 +972,34 @@ class elFinderVolumeFTP extends elFinderVolumeDriver {
[C,php/elFinderVolumeFTP.class.php,972] 		return false;
[C,php/elFinderVolumeFTP.class.php,973] 	}
[C,php/elFinderVolumeFTP.class.php,974] 	
[C,php/elFinderVolumeFTP.class.php,975] 	/**
[C,php/elFinderVolumeFTP.class.php,976] 	 * Return object width and height
[C,php/elFinderVolumeFTP.class.php,977] 	 * Ususaly used for images, but can be realize for video etc...
[C,php/elFinderVolumeFTP.class.php,978] 	 *
[C,php/elFinderVolumeFTP.class.php,979] 	 * @param  string  $path  file path
[C,php/elFinderVolumeFTP.class.php,980] 	 * @param  string  $mime  file mime type
[C,php/elFinderVolumeFTP.class.php,981] 	 * @return string|false
[C,php/elFinderVolumeFTP.class.php,982] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeFTP.class.php,983] 	 **/
[C,php/elFinderVolumeFTP.class.php,984] 	protected function _dimensions($path, $mime) {
[C,php/elFinderVolumeFTP.class.php,985] 		$ret = false;
[C,php/elFinderVolumeFTP.class.php,986] 		if ($imgsize = $this->getImageSize($path, $mime)) {
[D,php/elFinderVolumeFTP.class.php,987] 			$ret = $imgsize['dimensions'];
[A,php/elFinderVolumeFTP.class.php,987] 			$ret = array('dim' => $imgsize['dimensions']);
[A,php/elFinderVolumeFTP.class.php,988] 			if (!empty($imgsize['url'])) {
[A,php/elFinderVolumeFTP.class.php,989] 				$ret['url'] = $imgsize['url'];
[A,php/elFinderVolumeFTP.class.php,990] 			}
[C,php/elFinderVolumeFTP.class.php,991] 		}
[C,php/elFinderVolumeFTP.class.php,992] 		return $ret;
[C,php/elFinderVolumeFTP.class.php,993] 	}
[C,php/elFinderVolumeFTP.class.php,994] 	
[C,php/elFinderVolumeFTP.class.php,995] 	/******************** file/dir content *********************/
[C,php/elFinderVolumeFTP.class.php,996] 		
[C,php/elFinderVolumeFTP.class.php,997] 	/**
[C,php/elFinderVolumeFTP.class.php,998] 	 * Return files list in directory.
[C,php/elFinderVolumeFTP.class.php,999] 	 *
[C,php/elFinderVolumeFTP.class.php,1000] 	 * @param  string  $path  dir path
[C,php/elFinderVolumeFTP.class.php,1001] 	 * @return array
[C,php/elFinderVolumeFTP.class.php,1002] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeFTP.class.php,1003] 	 * @author Cem (DiscoFever)
[C,php/elFinderVolumeFTP.class.php,1004] 	 **/
[C,php/elFinderVolumeFTP.class.php,1005] 	protected function _scandir($path) {
@@ -1555,31 +1555,48 @@ class elFinderVolumeGoogleDrive extends elFinderVolumeDriver
[C,php/elFinderVolumeGoogleDrive.class.php,1555]      * @param string $mime file mime type
[C,php/elFinderVolumeGoogleDrive.class.php,1556]      *
[C,php/elFinderVolumeGoogleDrive.class.php,1557]      * @return string
[C,php/elFinderVolumeGoogleDrive.class.php,1558]      *
[C,php/elFinderVolumeGoogleDrive.class.php,1559]      * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeGoogleDrive.class.php,1560]      **/
[C,php/elFinderVolumeGoogleDrive.class.php,1561]     protected function _dimensions($path, $mime)
[C,php/elFinderVolumeGoogleDrive.class.php,1562]     {
[C,php/elFinderVolumeGoogleDrive.class.php,1563]         if (strpos($mime, 'image') !== 0) {
[C,php/elFinderVolumeGoogleDrive.class.php,1564]             return '';
[C,php/elFinderVolumeGoogleDrive.class.php,1565]         }
[C,php/elFinderVolumeGoogleDrive.class.php,1566]         $ret = '';
[C,php/elFinderVolumeGoogleDrive.class.php,1567] 
[C,php/elFinderVolumeGoogleDrive.class.php,1568]         if ($file = $this->_gd_getFile($path)) {
[C,php/elFinderVolumeGoogleDrive.class.php,1569]             if (isset($file['imageMediaMetadata'])) {
[D,php/elFinderVolumeGoogleDrive.class.php,1570]                 return $file['imageMediaMetadata']['width'].'x'.$file['imageMediaMetadata']['height'];
[A,php/elFinderVolumeGoogleDrive.class.php,1570]                 $ret = array('dim' => $file['imageMediaMetadata']['width'].'x'.$file['imageMediaMetadata']['height']);
[A,php/elFinderVolumeGoogleDrive.class.php,1571]                 if (func_num_args() > 2) {
[A,php/elFinderVolumeGoogleDrive.class.php,1572]                     $args = func_get_arg(2);
[A,php/elFinderVolumeGoogleDrive.class.php,1573]                 } else {
[A,php/elFinderVolumeGoogleDrive.class.php,1574]                     $args = array();
[A,php/elFinderVolumeGoogleDrive.class.php,1575]                 }
[A,php/elFinderVolumeGoogleDrive.class.php,1576]                 if (!empty($args['substitute'])) {
[A,php/elFinderVolumeGoogleDrive.class.php,1577]                     $tmbSize = intval($args['substitute']);
[A,php/elFinderVolumeGoogleDrive.class.php,1578]                     $srcSize = explode('x', $ret['dim']);
[A,php/elFinderVolumeGoogleDrive.class.php,1579]                     if (min(($tmbSize / $srcSize[0]), ($tmbSize / $srcSize[1])) < 1) {
[A,php/elFinderVolumeGoogleDrive.class.php,1580]                         if ($this->_gd_isPublished($file)) {
[A,php/elFinderVolumeGoogleDrive.class.php,1581]                     		$tmbSize = strval($tmbSize);
[A,php/elFinderVolumeGoogleDrive.class.php,1582]                         	$ret['url'] = 'https://drive.google.com/thumbnail?authuser=0&sz=s'.$tmbSize.'&id='.$file['id'];
[A,php/elFinderVolumeGoogleDrive.class.php,1583]                         } else if ($subImgLink = $this->getSubstituteImgLink(elFinder::$currentArgs['target'], $srcSize)) {
[A,php/elFinderVolumeGoogleDrive.class.php,1584]                         	$ret['url'] = $subImgLink;
[A,php/elFinderVolumeGoogleDrive.class.php,1585]                         }
[A,php/elFinderVolumeGoogleDrive.class.php,1586]                     }
[A,php/elFinderVolumeGoogleDrive.class.php,1587]                 }
[C,php/elFinderVolumeGoogleDrive.class.php,1588]             }
[C,php/elFinderVolumeGoogleDrive.class.php,1589]         }
[C,php/elFinderVolumeGoogleDrive.class.php,1590] 
[C,php/elFinderVolumeGoogleDrive.class.php,1591]         return $ret;
[C,php/elFinderVolumeGoogleDrive.class.php,1592]     }
[C,php/elFinderVolumeGoogleDrive.class.php,1593] 
[C,php/elFinderVolumeGoogleDrive.class.php,1594]     /******************** file/dir content *********************/
[C,php/elFinderVolumeGoogleDrive.class.php,1595] 
[C,php/elFinderVolumeGoogleDrive.class.php,1596]     /**
[C,php/elFinderVolumeGoogleDrive.class.php,1597]      * Return files list in directory.
[C,php/elFinderVolumeGoogleDrive.class.php,1598]      *
[C,php/elFinderVolumeGoogleDrive.class.php,1599]      * @param string $path dir path
[C,php/elFinderVolumeGoogleDrive.class.php,1600]      *
[C,php/elFinderVolumeGoogleDrive.class.php,1601]      * @return array
[C,php/elFinderVolumeGoogleDrive.class.php,1602]      *
@@ -1923,31 +1940,31 @@ class elFinderVolumeGoogleDrive extends elFinderVolumeDriver
[C,php/elFinderVolumeGoogleDrive.class.php,1940]                         'fields' => self::FETCHFIELDS_GET,
[C,php/elFinderVolumeGoogleDrive.class.php,1941]                     ]);
[C,php/elFinderVolumeGoogleDrive.class.php,1942]                 } else {
[C,php/elFinderVolumeGoogleDrive.class.php,1943]                     $request = $this->service->files->update($srcFile->getId(), $file, [
[C,php/elFinderVolumeGoogleDrive.class.php,1944]                         'fields' => self::FETCHFIELDS_GET,
[C,php/elFinderVolumeGoogleDrive.class.php,1945]                     ]);
[C,php/elFinderVolumeGoogleDrive.class.php,1946]                 }
[C,php/elFinderVolumeGoogleDrive.class.php,1947] 
[C,php/elFinderVolumeGoogleDrive.class.php,1948]                 // Create a media file upload to represent our upload process.
[C,php/elFinderVolumeGoogleDrive.class.php,1949]                 $media = new Google_Http_MediaFileUpload($client, $request, $mime, null, true, $chunkSizeBytes);
[C,php/elFinderVolumeGoogleDrive.class.php,1950]                 $media->setFileSize($size);
[C,php/elFinderVolumeGoogleDrive.class.php,1951]                 // Upload the various chunks. $status will be false until the process is
[C,php/elFinderVolumeGoogleDrive.class.php,1952]                 // complete.
[C,php/elFinderVolumeGoogleDrive.class.php,1953]                 $status = false;
[C,php/elFinderVolumeGoogleDrive.class.php,1954]                 while (!$status && !feof($fp)) {
[D,php/elFinderVolumeGoogleDrive.class.php,1938]                     elFinder::extendTimeLimit();
[A,php/elFinderVolumeGoogleDrive.class.php,1955]                     elFinder::checkAborted();
[C,php/elFinderVolumeGoogleDrive.class.php,1956]                     // read until you get $chunkSizeBytes from TESTFILE
[C,php/elFinderVolumeGoogleDrive.class.php,1957]                     // fread will never return more than 8192 bytes if the stream is read buffered and it does not represent a plain file
[C,php/elFinderVolumeGoogleDrive.class.php,1958]                     // An example of a read buffered file is when reading from a URL
[C,php/elFinderVolumeGoogleDrive.class.php,1959]                     $chunk = $this->_gd_readFileChunk($fp, $chunkSizeBytes);
[C,php/elFinderVolumeGoogleDrive.class.php,1960]                     $status = $media->nextChunk($chunk);
[C,php/elFinderVolumeGoogleDrive.class.php,1961]                 }
[C,php/elFinderVolumeGoogleDrive.class.php,1962]                 // The final value of $status will be the data from the API for the object
[C,php/elFinderVolumeGoogleDrive.class.php,1963]                 // that has been uploaded.
[C,php/elFinderVolumeGoogleDrive.class.php,1964]                 if ($status !== false) {
[C,php/elFinderVolumeGoogleDrive.class.php,1965]                     $obj = $status;
[C,php/elFinderVolumeGoogleDrive.class.php,1966]                 }
[C,php/elFinderVolumeGoogleDrive.class.php,1967] 
[C,php/elFinderVolumeGoogleDrive.class.php,1968]                 $client->setDefer(false);
[C,php/elFinderVolumeGoogleDrive.class.php,1969]             } else {
[C,php/elFinderVolumeGoogleDrive.class.php,1970]                 $params = [
@@ -47,30 +47,31 @@ class elFinderVolumeLocalFileSystem extends elFinderVolumeDriver {
[C,php/elFinderVolumeLocalFileSystem.class.php,47] 	/**
[C,php/elFinderVolumeLocalFileSystem.class.php,48] 	 * Constructor
[C,php/elFinderVolumeLocalFileSystem.class.php,49] 	 * Extend options with required fields
[C,php/elFinderVolumeLocalFileSystem.class.php,50] 	 *
[C,php/elFinderVolumeLocalFileSystem.class.php,51] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeLocalFileSystem.class.php,52] 	 */
[C,php/elFinderVolumeLocalFileSystem.class.php,53] 	public function __construct() {
[C,php/elFinderVolumeLocalFileSystem.class.php,54] 		$this->options['alias']    = '';              // alias to replace root dir name
[C,php/elFinderVolumeLocalFileSystem.class.php,55] 		$this->options['dirMode']  = 0755;            // new dirs mode
[C,php/elFinderVolumeLocalFileSystem.class.php,56] 		$this->options['fileMode'] = 0644;            // new files mode
[C,php/elFinderVolumeLocalFileSystem.class.php,57] 		$this->options['quarantine'] = '.quarantine'; // quarantine folder name - required to check archive (must be hidden)
[C,php/elFinderVolumeLocalFileSystem.class.php,58] 		$this->options['rootCssClass'] = 'elfinder-navbar-root-local';
[C,php/elFinderVolumeLocalFileSystem.class.php,59] 		$this->options['followSymLinks'] = true;
[C,php/elFinderVolumeLocalFileSystem.class.php,60] 		$this->options['detectDirIcon'] = '';         // file name that is detected as a folder icon e.g. '.diricon.png'
[C,php/elFinderVolumeLocalFileSystem.class.php,61] 		$this->options['keepTimestamp'] = array('copy', 'move'); // keep timestamp at inner filesystem allowed 'copy', 'move' and 'upload'
[A,php/elFinderVolumeLocalFileSystem.class.php,62] 		$this->options['substituteImg'] = true;       // support substitute image with dim command
[C,php/elFinderVolumeLocalFileSystem.class.php,63] 	}
[C,php/elFinderVolumeLocalFileSystem.class.php,64] 	
[C,php/elFinderVolumeLocalFileSystem.class.php,65] 	/*********************************************************************/
[C,php/elFinderVolumeLocalFileSystem.class.php,66] 	/*                        INIT AND CONFIGURE                         */
[C,php/elFinderVolumeLocalFileSystem.class.php,67] 	/*********************************************************************/
[C,php/elFinderVolumeLocalFileSystem.class.php,68] 	
[C,php/elFinderVolumeLocalFileSystem.class.php,69] 	/**
[C,php/elFinderVolumeLocalFileSystem.class.php,70] 	 * Prepare driver before mount volume.
[C,php/elFinderVolumeLocalFileSystem.class.php,71] 	 * Return true if volume is ready.
[C,php/elFinderVolumeLocalFileSystem.class.php,72] 	 *
[C,php/elFinderVolumeLocalFileSystem.class.php,73] 	 * @return bool
[C,php/elFinderVolumeLocalFileSystem.class.php,74] 	 **/
[C,php/elFinderVolumeLocalFileSystem.class.php,75] 	protected function init() {
[C,php/elFinderVolumeLocalFileSystem.class.php,76] 		// Normalize directory separator for windows
[C,php/elFinderVolumeLocalFileSystem.class.php,77] 		if (DIRECTORY_SEPARATOR !== '/') {
@@ -632,31 +633,31 @@ class elFinderVolumeLocalFileSystem extends elFinderVolumeDriver {
[C,php/elFinderVolumeLocalFileSystem.class.php,633] 		if (!file_exists($target)) {
[C,php/elFinderVolumeLocalFileSystem.class.php,634] 			return false;
[C,php/elFinderVolumeLocalFileSystem.class.php,635] 		}
[C,php/elFinderVolumeLocalFileSystem.class.php,636] 		
[C,php/elFinderVolumeLocalFileSystem.class.php,637] 		return $target;
[C,php/elFinderVolumeLocalFileSystem.class.php,638] 	}
[C,php/elFinderVolumeLocalFileSystem.class.php,639] 		
[C,php/elFinderVolumeLocalFileSystem.class.php,640] 	/**
[C,php/elFinderVolumeLocalFileSystem.class.php,641] 	 * Return files list in directory.
[C,php/elFinderVolumeLocalFileSystem.class.php,642] 	 *
[C,php/elFinderVolumeLocalFileSystem.class.php,643] 	 * @param  string  $path  dir path
[C,php/elFinderVolumeLocalFileSystem.class.php,644] 	 * @return array
[C,php/elFinderVolumeLocalFileSystem.class.php,645] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeLocalFileSystem.class.php,646] 	 **/
[C,php/elFinderVolumeLocalFileSystem.class.php,647] 	protected function _scandir($path) {
[D,php/elFinderVolumeLocalFileSystem.class.php,647] 		elFinder::extendTimeLimit();
[A,php/elFinderVolumeLocalFileSystem.class.php,648] 		elFinder::checkAborted();
[C,php/elFinderVolumeLocalFileSystem.class.php,649] 		$files = array();
[C,php/elFinderVolumeLocalFileSystem.class.php,650] 		$cache = array();
[C,php/elFinderVolumeLocalFileSystem.class.php,651] 		$dirWritable = is_writable($path);
[C,php/elFinderVolumeLocalFileSystem.class.php,652] 		$statOwner = (!empty($this->options['statOwner']));
[C,php/elFinderVolumeLocalFileSystem.class.php,653] 		$dirItr = array();
[C,php/elFinderVolumeLocalFileSystem.class.php,654] 		$followSymLinks = $this->options['followSymLinks'];
[C,php/elFinderVolumeLocalFileSystem.class.php,655] 		try {
[C,php/elFinderVolumeLocalFileSystem.class.php,656] 			$dirItr = new DirectoryIterator($path);
[C,php/elFinderVolumeLocalFileSystem.class.php,657] 		} catch (UnexpectedValueException $e) {}
[C,php/elFinderVolumeLocalFileSystem.class.php,658] 		
[C,php/elFinderVolumeLocalFileSystem.class.php,659] 		foreach ($dirItr as $file) {
[C,php/elFinderVolumeLocalFileSystem.class.php,660] 			try {
[C,php/elFinderVolumeLocalFileSystem.class.php,661] 				if ($file->isDot()) { continue; }
[C,php/elFinderVolumeLocalFileSystem.class.php,662] 				
[C,php/elFinderVolumeLocalFileSystem.class.php,663] 				$files[] = $fpath = $file->getPathname();
@@ -619,33 +619,34 @@ class elFinderVolumeMySQL extends elFinderVolumeDriver {
[C,php/elFinderVolumeMySQL.class.php,619] 			}
[C,php/elFinderVolumeMySQL.class.php,620] 		}
[C,php/elFinderVolumeMySQL.class.php,621] 		
[C,php/elFinderVolumeMySQL.class.php,622] 		return false;
[C,php/elFinderVolumeMySQL.class.php,623] 	}
[C,php/elFinderVolumeMySQL.class.php,624] 
[C,php/elFinderVolumeMySQL.class.php,625] 	/**
[C,php/elFinderVolumeMySQL.class.php,626] 	 * Close opened file
[C,php/elFinderVolumeMySQL.class.php,627] 	 *
[C,php/elFinderVolumeMySQL.class.php,628] 	 * @param  resource $fp file pointer
[C,php/elFinderVolumeMySQL.class.php,629] 	 * @param string $path
[C,php/elFinderVolumeMySQL.class.php,630] 	 * @return bool
[C,php/elFinderVolumeMySQL.class.php,631] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeMySQL.class.php,632] 	 */
[C,php/elFinderVolumeMySQL.class.php,633] 	protected function _fclose($fp, $path='') {
[D,php/elFinderVolumeMySQL.class.php,634] 		fclose($fp);
[A,php/elFinderVolumeMySQL.class.php,634] 		is_resource($fp) && fclose($fp);
[C,php/elFinderVolumeMySQL.class.php,635] 		if ($path) {
[D,php/elFinderVolumeMySQL.class.php,636] 			unlink($this->getTempFile($path));
[A,php/elFinderVolumeMySQL.class.php,636] 			$file = $this->getTempFile($path);
[A,php/elFinderVolumeMySQL.class.php,637] 			is_file($file) && unlink($file);
[C,php/elFinderVolumeMySQL.class.php,638] 		}
[C,php/elFinderVolumeMySQL.class.php,639] 	}
[C,php/elFinderVolumeMySQL.class.php,640] 	
[C,php/elFinderVolumeMySQL.class.php,641] 	/********************  file/dir manipulations *************************/
[C,php/elFinderVolumeMySQL.class.php,642] 	
[C,php/elFinderVolumeMySQL.class.php,643] 	/**
[C,php/elFinderVolumeMySQL.class.php,644] 	 * Create dir and return created dir path or false on failed
[C,php/elFinderVolumeMySQL.class.php,645] 	 *
[C,php/elFinderVolumeMySQL.class.php,646] 	 * @param  string  $path  parent dir path
[C,php/elFinderVolumeMySQL.class.php,647] 	 * @param string  $name  new directory name
[C,php/elFinderVolumeMySQL.class.php,648] 	 * @return string|bool
[C,php/elFinderVolumeMySQL.class.php,649] 	 * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeMySQL.class.php,650] 	 **/
[C,php/elFinderVolumeMySQL.class.php,651] 	protected function _mkdir($path, $name) {
[C,php/elFinderVolumeMySQL.class.php,652] 		return $this->make($path, $name, 'directory') ? $this->_joinPath($path, $name) : false;
@@ -311,31 +311,31 @@ class elFinderVolumeOneDrive extends elFinderVolumeDriver
[C,php/elFinderVolumeOneDrive.class.php,311] 
[C,php/elFinderVolumeOneDrive.class.php,312]         curl_setopt_array($curl, $defaultOptions);
[C,php/elFinderVolumeOneDrive.class.php,313] 
[C,php/elFinderVolumeOneDrive.class.php,314]         return $curl;
[C,php/elFinderVolumeOneDrive.class.php,315]     }
[C,php/elFinderVolumeOneDrive.class.php,316] 
[C,php/elFinderVolumeOneDrive.class.php,317]     /**
[C,php/elFinderVolumeOneDrive.class.php,318]      * Creates a base cURL object which is compatible with the OneDrive API.
[C,php/elFinderVolumeOneDrive.class.php,319]      *
[C,php/elFinderVolumeOneDrive.class.php,320]      * @param string $path The path of the API call (eg. me/skydrive)
[C,php/elFinderVolumeOneDrive.class.php,321]      *
[C,php/elFinderVolumeOneDrive.class.php,322]      * @return resource A compatible cURL object
[C,php/elFinderVolumeOneDrive.class.php,323]      */
[C,php/elFinderVolumeOneDrive.class.php,324]     protected function _od_createCurl($path, $contents = false)
[C,php/elFinderVolumeOneDrive.class.php,325]     {
[D,php/elFinderVolumeOneDrive.class.php,326]         elfinder::extendTimeLimit();
[A,php/elFinderVolumeOneDrive.class.php,326]         elFinder::checkAborted();
[C,php/elFinderVolumeOneDrive.class.php,327]         $curl = $this->_od_prepareCurl($path);
[C,php/elFinderVolumeOneDrive.class.php,328] 
[C,php/elFinderVolumeOneDrive.class.php,329]         if ($contents) {
[C,php/elFinderVolumeOneDrive.class.php,330]             $res = elFinder::curlExec($curl);
[C,php/elFinderVolumeOneDrive.class.php,331]         } else {
[C,php/elFinderVolumeOneDrive.class.php,332]             $result = json_decode(curl_exec($curl));
[C,php/elFinderVolumeOneDrive.class.php,333]             curl_close($curl);
[C,php/elFinderVolumeOneDrive.class.php,334]             if (isset($result->value)) {
[C,php/elFinderVolumeOneDrive.class.php,335]                 $res = $result->value;
[C,php/elFinderVolumeOneDrive.class.php,336]                 unset($result->value);
[C,php/elFinderVolumeOneDrive.class.php,337]                 $result = (array) $result;
[C,php/elFinderVolumeOneDrive.class.php,338]                 if (!empty($result['@odata.nextLink'])) {
[C,php/elFinderVolumeOneDrive.class.php,339]                     $nextRes = $this->_od_createCurl($result['@odata.nextLink'], false);
[C,php/elFinderVolumeOneDrive.class.php,340]                     if (is_array($nextRes)) {
[C,php/elFinderVolumeOneDrive.class.php,341]                         $res = array_merge($res, $nextRes);
@@ -673,46 +673,46 @@ class elFinderVolumeOneDrive extends elFinderVolumeDriver
[C,php/elFinderVolumeOneDrive.class.php,673]      *
[C,php/elFinderVolumeOneDrive.class.php,674]      * @param array $options
[C,php/elFinderVolumeOneDrive.class.php,675]      *
[C,php/elFinderVolumeOneDrive.class.php,676]      * @return bool
[C,php/elFinderVolumeOneDrive.class.php,677]      */
[C,php/elFinderVolumeOneDrive.class.php,678]     public function netunmount($netVolumes, $key)
[C,php/elFinderVolumeOneDrive.class.php,679]     {
[C,php/elFinderVolumeOneDrive.class.php,680]         if ($tmbs = glob(rtrim($this->options['tmbPath'], '\\/').DIRECTORY_SEPARATOR.$this->tmbPrefix.'*.png')) {
[C,php/elFinderVolumeOneDrive.class.php,681]             foreach ($tmbs as $file) {
[C,php/elFinderVolumeOneDrive.class.php,682]                 unlink($file);
[C,php/elFinderVolumeOneDrive.class.php,683]             }
[C,php/elFinderVolumeOneDrive.class.php,684]         }
[C,php/elFinderVolumeOneDrive.class.php,685] 
[C,php/elFinderVolumeOneDrive.class.php,686]         return true;
[C,php/elFinderVolumeOneDrive.class.php,687]     }
[D,php/elFinderVolumeOneDrive.class.php,688]     
[A,php/elFinderVolumeOneDrive.class.php,688] 
[C,php/elFinderVolumeOneDrive.class.php,689]     /**
[C,php/elFinderVolumeOneDrive.class.php,690]      * Return debug info for client.
[C,php/elFinderVolumeOneDrive.class.php,691]      *
[C,php/elFinderVolumeOneDrive.class.php,692]      * @return array
[C,php/elFinderVolumeOneDrive.class.php,693]      **/
[C,php/elFinderVolumeOneDrive.class.php,694]     public function debug()
[C,php/elFinderVolumeOneDrive.class.php,695]     {
[C,php/elFinderVolumeOneDrive.class.php,696]         $res = parent::debug();
[D,php/elFinderVolumeOneDrive.class.php,697]         if (! empty($this->options['accessToken'])) {
[A,php/elFinderVolumeOneDrive.class.php,697]         if (!empty($this->options['accessToken'])) {
[C,php/elFinderVolumeOneDrive.class.php,698]             $res['accessToken'] = $this->options['accessToken'];
[C,php/elFinderVolumeOneDrive.class.php,699]         }
[D,php/elFinderVolumeOneDrive.class.php,700]     
[A,php/elFinderVolumeOneDrive.class.php,700] 
[C,php/elFinderVolumeOneDrive.class.php,701]         return $res;
[C,php/elFinderVolumeOneDrive.class.php,702]     }
[D,php/elFinderVolumeOneDrive.class.php,703]     
[A,php/elFinderVolumeOneDrive.class.php,703] 
[C,php/elFinderVolumeOneDrive.class.php,704]     /*********************************************************************/
[C,php/elFinderVolumeOneDrive.class.php,705]     /*                        INIT AND CONFIGURE                         */
[C,php/elFinderVolumeOneDrive.class.php,706]     /*********************************************************************/
[C,php/elFinderVolumeOneDrive.class.php,707] 
[C,php/elFinderVolumeOneDrive.class.php,708]     /**
[C,php/elFinderVolumeOneDrive.class.php,709]      * Prepare FTP connection
[C,php/elFinderVolumeOneDrive.class.php,710]      * Connect to remote server and check if credentials are correct, if so, store the connection id in $ftp_conn.
[C,php/elFinderVolumeOneDrive.class.php,711]      *
[C,php/elFinderVolumeOneDrive.class.php,712]      * @return bool
[C,php/elFinderVolumeOneDrive.class.php,713]      *
[C,php/elFinderVolumeOneDrive.class.php,714]      * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeOneDrive.class.php,715]      * @author Cem (DiscoFever)
[C,php/elFinderVolumeOneDrive.class.php,716]      **/
[C,php/elFinderVolumeOneDrive.class.php,717]     protected function init()
[C,php/elFinderVolumeOneDrive.class.php,718]     {
@@ -1255,35 +1255,69 @@ class elFinderVolumeOneDrive extends elFinderVolumeDriver
[C,php/elFinderVolumeOneDrive.class.php,1255]      * Ususaly used for images, but can be realize for video etc...
[C,php/elFinderVolumeOneDrive.class.php,1256]      *
[C,php/elFinderVolumeOneDrive.class.php,1257]      * @param string $path file path
[C,php/elFinderVolumeOneDrive.class.php,1258]      * @param string $mime file mime type
[C,php/elFinderVolumeOneDrive.class.php,1259]      *
[C,php/elFinderVolumeOneDrive.class.php,1260]      * @return string
[C,php/elFinderVolumeOneDrive.class.php,1261]      *
[C,php/elFinderVolumeOneDrive.class.php,1262]      * @author Dmitry (dio) Levashov
[C,php/elFinderVolumeOneDrive.class.php,1263]      **/
[C,php/elFinderVolumeOneDrive.class.php,1264]     protected function _dimensions($path, $mime)
[C,php/elFinderVolumeOneDrive.class.php,1265]     {
[C,php/elFinderVolumeOneDrive.class.php,1266]         if (strpos($mime, 'image') !== 0) {
[C,php/elFinderVolumeOneDrive.class.php,1267]             return '';
[C,php/elFinderVolumeOneDrive.class.php,1268]         }
[C,php/elFinderVolumeOneDrive.class.php,1269] 
[D,php/elFinderVolumeOneDrive.class.php,1270]         $cache = $this->_od_getFileRaw($path);
[A,php/elFinderVolumeOneDrive.class.php,1270]         //$cache = $this->_od_getFileRaw($path);
[A,php/elFinderVolumeOneDrive.class.php,1271]         if (func_num_args() > 2) {
[A,php/elFinderVolumeOneDrive.class.php,1272]             $args = func_get_arg(2);
[A,php/elFinderVolumeOneDrive.class.php,1273]         } else {
[A,php/elFinderVolumeOneDrive.class.php,1274]             $args = array();
[A,php/elFinderVolumeOneDrive.class.php,1275]         }
[A,php/elFinderVolumeOneDrive.class.php,1276]         if (!empty($args['substitute'])) {
[A,php/elFinderVolumeOneDrive.class.php,1277]             $tmbSize = intval($args['substitute']);
[A,php/elFinderVolumeOneDrive.class.php,1278]         } else {
[A,php/elFinderVolumeOneDrive.class.php,1279]             $tmbSize = null;
[A,php/elFinderVolumeOneDrive.class.php,1280]         }
[A,php/elFinderVolumeOneDrive.class.php,1281]         list(, $itemId) = $this->_od_splitPath($path);
[A,php/elFinderVolumeOneDrive.class.php,1282]         $options = array(
[A,php/elFinderVolumeOneDrive.class.php,1283]             'query' => array(
[A,php/elFinderVolumeOneDrive.class.php,1284]                 'select' => 'id,image',
[A,php/elFinderVolumeOneDrive.class.php,1285]             ),
[A,php/elFinderVolumeOneDrive.class.php,1286]         );
[A,php/elFinderVolumeOneDrive.class.php,1287]         if ($tmbSize) {
[A,php/elFinderVolumeOneDrive.class.php,1288]             $tmb = 'c'.$tmbSize.'x'.$tmbSize;
[A,php/elFinderVolumeOneDrive.class.php,1289]             $options['query']['expand'] = 'thumbnails(select='.$tmb.')';
[A,php/elFinderVolumeOneDrive.class.php,1290]         }
[A,php/elFinderVolumeOneDrive.class.php,1291]         $raw = $this->_od_query($itemId, true, false, $options);
[C,php/elFinderVolumeOneDrive.class.php,1292] 
[D,php/elFinderVolumeOneDrive.class.php,1272]         if ($cache && $img = $cache->image) {
[A,php/elFinderVolumeOneDrive.class.php,1293]         if ($raw && $img = $raw->image) {
[C,php/elFinderVolumeOneDrive.class.php,1294]             if (isset($img->width) && isset($img->height)) {
[D,php/elFinderVolumeOneDrive.class.php,1274]                 return $img->width.'x'.$img->height;
[A,php/elFinderVolumeOneDrive.class.php,1295]                 $ret = array('dim' => $img->width.'x'.$img->height);
[A,php/elFinderVolumeOneDrive.class.php,1296]                 if ($tmbSize) {
[A,php/elFinderVolumeOneDrive.class.php,1297]                     $srcSize = explode('x', $ret['dim']);
[A,php/elFinderVolumeOneDrive.class.php,1298]                     if (min(($tmbSize / $srcSize[0]), ($tmbSize / $srcSize[1])) < 1) {
[A,php/elFinderVolumeOneDrive.class.php,1299]                         if (!empty($raw->thumbnails)) {
[A,php/elFinderVolumeOneDrive.class.php,1300]                             $tmbArr = (array) $raw->thumbnails[0];
[A,php/elFinderVolumeOneDrive.class.php,1301]                             if (!empty($tmbArr[$tmb]->url)) {
[A,php/elFinderVolumeOneDrive.class.php,1302]                                 $ret['url'] = $tmbArr[$tmb]->url;
[A,php/elFinderVolumeOneDrive.class.php,1303]                             }
[A,php/elFinderVolumeOneDrive.class.php,1304]                         }
[A,php/elFinderVolumeOneDrive.class.php,1305]                     }
[A,php/elFinderVolumeOneDrive.class.php,1306]                 }
[A,php/elFinderVolumeOneDrive.class.php,1307] 
[A,php/elFinderVolumeOneDrive.class.php,1308]                 return $ret;
[C,php/elFinderVolumeOneDrive.class.php,1309]             }
[C,php/elFinderVolumeOneDrive.class.php,1310]         }
[C,php/elFinderVolumeOneDrive.class.php,1311] 
[C,php/elFinderVolumeOneDrive.class.php,1312]         $ret = '';
[C,php/elFinderVolumeOneDrive.class.php,1313]         if ($work = $this->getWorkFile($path)) {
[C,php/elFinderVolumeOneDrive.class.php,1314]             if ($size = @getimagesize($work)) {
[C,php/elFinderVolumeOneDrive.class.php,1315]                 $cache['width'] = $size[0];
[C,php/elFinderVolumeOneDrive.class.php,1316]                 $cache['height'] = $size[1];
[C,php/elFinderVolumeOneDrive.class.php,1317]                 $ret = $size[0].'x'.$size[1];
[C,php/elFinderVolumeOneDrive.class.php,1318]             }
[C,php/elFinderVolumeOneDrive.class.php,1319]         }
[C,php/elFinderVolumeOneDrive.class.php,1320]         is_file($work) && @unlink($work);
[C,php/elFinderVolumeOneDrive.class.php,1321] 
[C,php/elFinderVolumeOneDrive.class.php,1322]         return $ret;
[C,php/elFinderVolumeOneDrive.class.php,1323]     }
@@ -1457,44 +1491,44 @@ class elFinderVolumeOneDrive extends elFinderVolumeDriver
[C,php/elFinderVolumeOneDrive.class.php,1491]             $data = (object) $properties;
[C,php/elFinderVolumeOneDrive.class.php,1492]             $curl = $this->_od_prepareCurl($url);
[C,php/elFinderVolumeOneDrive.class.php,1493]             curl_setopt_array($curl, array(
[C,php/elFinderVolumeOneDrive.class.php,1494]                 CURLOPT_POST => true,
[C,php/elFinderVolumeOneDrive.class.php,1495]                 CURLOPT_HEADER => true,
[C,php/elFinderVolumeOneDrive.class.php,1496]                 CURLOPT_HTTPHEADER => array(
[C,php/elFinderVolumeOneDrive.class.php,1497]                       'Content-Type: application/json',
[C,php/elFinderVolumeOneDrive.class.php,1498]                        'Authorization: Bearer '.$this->token->data->access_token,
[C,php/elFinderVolumeOneDrive.class.php,1499]                     'Prefer: respond-async',
[C,php/elFinderVolumeOneDrive.class.php,1500]                 ),
[C,php/elFinderVolumeOneDrive.class.php,1501]                 CURLOPT_POSTFIELDS => json_encode($data),
[C,php/elFinderVolumeOneDrive.class.php,1502]             ));
[C,php/elFinderVolumeOneDrive.class.php,1503]             $result = curl_exec($curl);
[C,php/elFinderVolumeOneDrive.class.php,1504]             curl_close($curl);
[C,php/elFinderVolumeOneDrive.class.php,1505] 
[D,php/elFinderVolumeOneDrive.class.php,1472]             $res = new stdClass;
[A,php/elFinderVolumeOneDrive.class.php,1506]             $res = new stdClass();
[C,php/elFinderVolumeOneDrive.class.php,1507]             if (preg_match('/Location: (.+)/', $result, $m)) {
[C,php/elFinderVolumeOneDrive.class.php,1508]                 $monUrl = trim($m[1]);
[C,php/elFinderVolumeOneDrive.class.php,1509]                 while ($res) {
[C,php/elFinderVolumeOneDrive.class.php,1510]                     usleep(200000);
[C,php/elFinderVolumeOneDrive.class.php,1511]                     $curl = $this->_od_prepareCurl($monUrl);
[C,php/elFinderVolumeOneDrive.class.php,1512]                     $res = json_decode(curl_exec($curl));
[C,php/elFinderVolumeOneDrive.class.php,1513]                     curl_close($curl);
[C,php/elFinderVolumeOneDrive.class.php,1514]                     if (isset($res->status)) {
[C,php/elFinderVolumeOneDrive.class.php,1515]                         if ($res->status === 'completed' || $res->status === 'failed') {
[C,php/elFinderVolumeOneDrive.class.php,1516]                             break;
[C,php/elFinderVolumeOneDrive.class.php,1517]                         }
[C,php/elFinderVolumeOneDrive.class.php,1518]                     }
[D,php/elFinderVolumeOneDrive.class.php,1485]                  }
[A,php/elFinderVolumeOneDrive.class.php,1519]                 }
[C,php/elFinderVolumeOneDrive.class.php,1520]             }
[C,php/elFinderVolumeOneDrive.class.php,1521] 
[C,php/elFinderVolumeOneDrive.class.php,1522]             if ($res && isset($res->resourceId)) {
[C,php/elFinderVolumeOneDrive.class.php,1523]                 if (isset($res->folder) && isset($this->sessionCache['subdirs'])) {
[C,php/elFinderVolumeOneDrive.class.php,1524]                     $this->sessionCache['subdirs'][$targetDir] = true;
[C,php/elFinderVolumeOneDrive.class.php,1525]                 }
[C,php/elFinderVolumeOneDrive.class.php,1526] 
[C,php/elFinderVolumeOneDrive.class.php,1527]                 return $this->_joinPath($targetDir, $res->resourceId);
[C,php/elFinderVolumeOneDrive.class.php,1528]             }
[C,php/elFinderVolumeOneDrive.class.php,1529] 
[C,php/elFinderVolumeOneDrive.class.php,1530]             return false;
[C,php/elFinderVolumeOneDrive.class.php,1531]         } catch (Exception $e) {
[C,php/elFinderVolumeOneDrive.class.php,1532]             return $this->setError('OneDrive error: '.$e->getMessage());
[C,php/elFinderVolumeOneDrive.class.php,1533]         }
[C,php/elFinderVolumeOneDrive.class.php,1534] 